### PR TITLE
Implemented the Leaderboard

### PR DESCRIPTION
**Pull request by 54657566656c**
**How it looks:**
image

**How it works:**
- exactly like etherscan. navigate via buttons
- addition: toggle the number of shown elements between 10, 25, 50 and 100 by clicking the middle button (1-10 on the image above)
- addition: is displaying all entries, not just the top 1000

**How its implemented:**
- base data is hard coded (snapshot from earlier today)
- updates are fetched on startup, if there are any, via the etherscan api. I suggest using a api-key

### CODE REVIEW

File: src/views/Leaderboard.vue [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+      </button>
+      <button class="btn btn-primary" @click="onLast" style="margin: 5px">
+        Last
+      </button>
+    </div>
+  </div>
+</template>
+
+<script>
+import holder from "../store/data/holder.json";
```
Comment (bwcgn) [inline src/views/Leaderboard.vue:109]: that should be rather a blockchain information than a static data set

File: src/views/Leaderboard.vue [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+        this.currentIndex = newIndex;
+      }
+    },
+    onLast() {
+      this.currentIndex = this.leaderboard.length - this.elementsToDisplay;
+    },
+    async fetchTransfers(startBlock) {
+      const address = "0x93a796b1e846567fe3577af7b7bb89f71680173a";
+      const topic = "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef";
+      const apikey = "REPLACE_WITH_YOUR_API_KEY";
```
Comment (bwcgn) [inline src/views/Leaderboard.vue]: 👀

File: src/views/Leaderboard.vue (hunk 1 above)
Comment (54657566656c) [reply src/views/Leaderboard.vue:109]: Yes, I know. But requesting all that data takes some time, at least the way I do it, so I created this "snapshot" to boost start everything. Is there a better way to request the blockchain data? A faster one?

Comment (geggleto) [reply src/views/Leaderboard.vue:109]: It needs to be a static set ser, it takes forever to query the information.

File: src/views/Leaderboard.vue [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+      const url = `https://api.etherscan.io/api?module=logs&action=getLogs&fromBlock=${startBlock}&toBlock=latest&address=${address}&topic0=${topic}&apikey=${apikey}`;
```
Comment (geggleto) [inline src/views/Leaderboard.vue]: can we not use the code that already exists in the Store? this already exists and doesnt require a stupid api key <.<

**Review by geggleto — CHANGES_REQUESTED** [review]
dont rely on etherscan, all the query code should exist in the repository already for reading from the blockchain

File: src/views/Leaderboard.vue (hunk 3 above)
Comment (54657566656c) [reply src/views/Leaderboard.vue]: yes. i reworked it. have a look if you have time

**Comment by 54657566656c** [comment]
I added ENS names for wallets, if available. Below are some points to consider:
- According to the [ENS Documentation](https://docs.ens.domains/dapp-developer-guide/resolving-names#reverse-resolution), Web3.js does not support reverse ENS lookup. Therefore I used etherjs instead. 
- ENS name lookups will be triggered for the currently visible holders. For example, when opening the leaderboard, the places 1-10 are visible, which means requests for the ENS names for addresses 1-10 will be triggered.
- I configured etherjs with an Infura project (using environment variables INFURA_PROJECT_ID and INFURA_PROJECT_SECRET), but there are other provider options available ([etherjs provider](https://docs.ethers.io/v5/api/providers/)).
- To reduce the load I refreshed the static dataset with the ENS names of the top 1000 holders (which currently includes everyone with more than two nfts). Also, already existing ENS names are not receiving updates.
- Whenever a new ENS name is found, it will be added to the corresponding leaderboard entry. This does not result in a view refresh. Example: Open leaderboard -> visit last 10 holders -> no ENS visible, but fetching for those is triggered -> fetching might be done sometime later, but the list will not change -> visit first 10 holders again (or trigger any other table change of your liking) -> go back to the last 10 holders -> ENS names will show up (if the fetch requests are already done)

**Review by geggleto — APPROVED** [review]

**Pull request merged**